python/ffsim/qiskit/sim.py

```diff
@@ -767,6 +800,73 @@ def _extract_x_gates(circuit: QuantumCircuit) -> tuple[list[int], QuantumCircuit
     return indices, remaining_circuit
 
 
+def _permute_bitstrings(
+    bits: np.ndarray, pairs: Sequence[tuple[int, int]]
+) -> np.ndarray:
+    """Return bitstrings with bits moved according to (src, dest) pairs."""
+    if not pairs:
+        return bits
+    mask = sum(1 << dest for _, dest in pairs)
+    permuted = bits.copy()
+    if mask:
+        permuted &= ~mask
+    for src, dest in pairs:
+        permuted |= ((bits >> src) & 1) << dest
+    return permuted
+
+
+def _apply_permutation_gate(
+    vec: np.ndarray,
+    qubit_indices: Sequence[int],
+    perm: Sequence[int],
+    *,
+    norb: int,
+    nelec: int | tuple[int, int],
+    dests: Sequence[int] | None = None,
+) -> np.ndarray:
+    if dests is None:
+        inverse_perm = np.empty_like(perm)
+        inverse_perm[perm] = np.arange(len(perm))
+        dests = [qubit_indices[i] for i in inverse_perm]
+
+    pairs = list(zip(qubit_indices, dests))
+    addresses = range(len(vec))
+    if isinstance(nelec, int):
+        # Spinless case
+        bitstrings = np.asarray(
+            states.addresses_to_strings(
+                addresses,
+                norb=norb,
+                nelec=nelec,
+                bitstring_type=BitstringType.INT,
+            )
+        )
+        permuted = _permute_bitstrings(bitstrings, pairs)
+        indices = states.strings_to_addresses(permuted, norb=norb, nelec=nelec)
+    else:
+        # Spinful case
+        strings_a, strings_b = states.addresses_to_strings(
+            addresses,
+            norb=norb,
+            nelec=nelec,
+            bitstring_type=BitstringType.INT,
+            concatenate=False,
+        )
+        alpha_bits = np.asarray(strings_a)
+        beta_bits = np.asarray(strings_b)
+        alpha_pairs = [(src, dest) for src, dest in pairs if src < norb]
+        beta_pairs = [(src - norb, dest - norb) for src, dest in pairs if src >= norb]
+
+        permuted_alpha = _permute_bitstrings(alpha_bits, alpha_pairs)
+        permuted_beta = _permute_bitstrings(beta_bits, beta_pairs)
+
+        combined = permuted_alpha + (permuted_beta << norb)
```

Comment:
Could this create integers greater than 64 bits? Either way, please add a test for a large number of bits. For example, you could create a circuit with 64 orbitals but only two electrons.